File: src/cpp/multithreading/semaphore.cpp
/**
 * Semaphore
 *
 * This example demonstrates a counting semaphore - a synchronization
 * primitive that controls access to a limited pool of resources.
 *
 * Key concepts:
 * - Semaphore has an internal counter
 * - acquire() decrements counter (blocks if counter is 0)
 * - release() increments counter (wakes one waiting thread)
 * - Unlike mutex, semaphore can be released by a different thread
 * - Useful for limiting concurrent access to a resource pool
 *
 * Note: C++20 provides std::counting_semaphore, but this shows the mechanism.
 */

#include <chrono>
#include <condition_variable>
#include <iostream>
#include <mutex>
#include <random>
#include <thread>
#include <vector>

class Semaphore {
public:
    explicit Semaphore(unsigned long initial_count)
        : count_(initial_count) {}

    void acquire() {
        std::unique_lock<std::mutex> lock(mutex_);
        cv_.wait(lock, [this] { return count_ > 0; });
        --count_;
    }

    void release() {
        {
            std::scoped_lock lock(mutex_);
            ++count_;
        }
        cv_.notify_one();
    }

private:
    std::mutex mutex_;
    std::condition_variable cv_;
    unsigned long count_;
};

// Simulate a shared printer with limited access (1 at a time)
Semaphore printer(1);

void office_worker(int id) {
    // Acquire access to the printer
    printer.acquire();

    std::cout << "Worker " << id << " is using the printer...\n";

    // Simulate printing (random 1-3 seconds)
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dist(1000, 3000);
    std::this_thread::sleep_for(std::chrono::milliseconds(dist(gen)));

    std::cout << "Worker " << id << " finished printing.\n";

    // Release the printer for others
    printer.release();
}

int main() {
    std::vector<std::thread> workers;

    for (int i = 0; i < 5; ++i) {
        workers.emplace_back(office_worker, i);
    }

    for (auto& w : workers) {
        w.join();
    }

    std::cout << "All workers finished.\n";
    return 0;
}

File: src/python/multithreading/semaphore.py
"""
Semaphore Example

This script demonstrates synchronization using a semaphore. Unlike a mutex
(which allows only one thread), a semaphore limits access to a resource to N
threads simultaneously.

Key Concepts:
- Lock is the usual tool for protecting one shared critical section
- Semaphore(N) is the usual tool for representing N identical resource slots
- Use a semaphore as a context manager so release() happens automatically

Good Practice:
- Prefer Lock over Semaphore(1) for ordinary mutual exclusion because it states
  the intent more clearly.
- Prefer Semaphore(N) when the resource really has capacity N, such as a small
  connection pool, a download limit, or a fixed number of hardware devices.
"""

import random
import time
from threading import Semaphore, Thread

MAX_CONNECTIONS = 3
connection_slots = Semaphore(MAX_CONNECTIONS)


def use_connection(worker_id):
    """Simulate work that needs one slot from a limited connection pool."""
    print(f"Worker {worker_id} is waiting for a connection slot")

    with connection_slots:
        print(f"Worker {worker_id} acquired a connection slot")
        time.sleep(random.uniform(0.5, 1.5))
        print(f"Worker {worker_id} released a connection slot")


def main():
    threads = [Thread(target=use_connection, args=(i + 1,)) for i in range(10)]

    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()

    print("All workers completed.")


if __name__ == "__main__":
    main()
